src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -69,11 +69,23 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
 
         public override bool TryRead(out ReadResult readResult)
         {
-            var result = _context.RequestBodyPipe.Reader.TryRead(out readResult);
-            _readResult = readResult;
-            CountBytesRead(readResult.Buffer.Length);
+            TryStart();
```

Comment:
@halter73 note: had to include this `Http2MessageBody` change for `TryStart`/`TryStop` to prevent test failures.

Reply:
It definitely feels better to call TryStart here. It does important stuff that isn't even timing related like checking that the request body size isn't larger than what's allowed.

What test failures were you seeing specifically.

Reply:
One example was the following though there were a couple different types
```
Error message
Moq.MockException : \nExpected invocation on the mock should never have been performed, but was 32 times: h => h.OnTimeout(It.IsAny<TimeoutReason>())\r\n\r\nConfigured setups: \r\nITimeoutHandler h => h.OnTimeout(It.IsAny<TimeoutReason>())\r\n\r\nPerformed invocations: \r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)\r\nITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)

Stack trace
   at Moq.Mock.VerifyCalls(Mock targetMock, InvocationShape expectation, LambdaExpression expression, Times times, String failMessage)
   at Moq.Mock.Verify[T](Mock`1 mock, Expression`1 expression, Times times, String failMessage)
   at Moq.Mock`1.Verify(Expression`1 expression, Times times)
   at Moq.Mock`1.Verify(Expression`1 expression, Func`1 times)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TimeoutTests.DATA_Received_TooSlowlyOnSecondStream_AbortsConnectionAfterNonAdditiveRateTimeout() in /_/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:line 775
--- End of stack trace from previous location where exception was thrown ---
```

Reply:
Ah. That makes sense. If you start reading a request body with TryRead, and then call ReadAsync, Kestrel will reset _readTimingBytesRead to 0 on the first ReadAsync call ignoring the bytes counted by TryRead.

I wonder how common using both TryRead and ReadAsync on the same RequestBody will be before we do it implicitly ReadAsyncInternal as part of this PR. We might want to backport this part of the change.

Reply:
Does `Http1ContentLengthMessageBody` need this? 

`Http1ChunkedEncodingMessageBody` has it; but ContentLength doesn't have the same pattern (`TryRead` earlier and `TryStop`)

Reply:
Ah; is correct `TryStop` is in `CreateReadResultFromConnectionReadResult`

Added isolated change for Http2 https://github.com/aspnet/AspNetCore/pull/13305